### ISSUE TITLE
[self-solved] Is there any way that I can make boundary smooth ?

### ISSUE THREAD

**Issue by hongminpark** [facebookresearch/segment-anything#356]
Hi, I am new to image segmentation. 
The result of mask's border isn't smooth. 
![F3D370FC-643B-41F8-9DC2-7133C208769B_1_105_c jpeg_0](https://github.com/facebookresearch/segment-anything/assets/21957275/a6dae71f-0122-4faf-81e9-840b4000d91f)

Is it right to use libraries like [pymatting](https://github.com/pymatting/pymatting) to make border smooth ?
Or are there any other solution? 
Do I have to train with bigger size of images? 
Welcome any experts help :)

**Comment by wensle** [comment]
I’m not an expert by any means, nor do I have that much experience with SAM, but to me this looks like it’s just a matter of applying antialiasing. The masks don’t have antialiasing applied to them, nor do I think they should, but your cutout would benefit from it it seems. Just apply antialiasing to the banana, imo.

**Comment by hongminpark** [comment]
@wensle Hey thanks for your reply :)
I forgot about this issue but I self-solved applying 'GaussianBlur' + 'Contrast' effect on binary mask! 

Steps
- Binary mask to Grayscale -> apply Gaussian Blur -> adjust Contrast

```python
from scipy.ndimage import gaussian_filter
from skimage import exposure

def binary_to_grayscale(binary_image):
    grayscale_image = binary_image.astype(np.uint8) * 255
    return grayscale_image
  
def apply_gaussian_blur(image, sigma=1):
    blurred_image = gaussian_filter(image, sigma=sigma)
    return blurred_image

def adjust_contrast(image, gamma=1.0):
    adjusted_image = exposure.adjust_gamma(image, gamma)
    return adjusted_image

# Assuming 'array' is your binary image
grayscale_mask = binary_to_grayscale(masks[0]['segmentation'])
# blur the mask
blurred_mask = apply_gaussian_blur(grayscale_mask, sigma=1.5) # Make mask a little bit blurry. This makes edge smooth
# make mask brighter
adjusted_mask = adjust_contrast(blurred_mask, gamma=20) # Lighten the blur. This makes the blurred edge to look more like natural.
```
Just for comparison!! I might should make a code repo for others :) 
#### gaussian blur(sigma=1.5)
![jennys_banana_cropped](https://github.com/facebookresearch/segment-anything/assets/21957275/a8c742f3-f0f9-4ffc-9dc7-48b41e7051bb)

#### gaussian blur(sigma=1.5) + exposure(gamma=10)
![jennys_banana_cropped_gamma10](https://github.com/facebookresearch/segment-anything/assets/21957275/106b41b2-8453-4d33-b005-8303b4559be9)

#### gaussian blur(sigma=1.5) + exposure(gamma=20)
![jennys_banana_cropped_gamma20](https://github.com/facebookresearch/segment-anything/assets/21957275/e31814ae-0f6f-4245-92f1-ac84a4965f5e)

**Comment by hongminpark** [comment]
@dhirajpatra 
Try this:
Apply gaussian blur to mask (m = `ann['segmentation']). This would generate slightly enlarged mask with blurred edge. Imagine your masked image with blurred outline. OpenCV has `gaussinBlur` function. Use it like this: `cv2.GaussianBlur(m, kernel_size, sigma)`. Adjust kernel_size and sigma to make solid outline.
Refer to cv2 document or ask chatgpt! I use chatgpt a lot :) Above code is all gpt helped code.

**Comment by dhirajpatra** [comment]
Thank you @hongminpark however I am looking for the border of each segmented mask area. In meta demo it mask segment for each obj in an image along with dark blue or else border of each segment. 
Howvery not able to get that border.

**Comment by hongminpark** [comment]
Oh. That's just UI trick! The masked image doesn't have any border. In demo, it only has masked result but generating border with css's drop-shadow effect. If you know html/css, you can inspect the code in browser.
<img width="1261" alt="image" src="https://github.com/facebookresearch/segment-anything/assets/21957275/1a2136f4-2c5d-4d5d-a1b8-7e5b6bd539b9">
<img width="1140" alt="image" src="https://github.com/facebookresearch/segment-anything/assets/21957275/f0853fd7-9bf7-4664-9378-bc31e15be925">

**Comment by dhirajpatra** [comment]
Oh thanks a lot I will try @hongminpark

**Comment by abhishek0093** [comment]
Sharing this in case it helps somebody. 
I tried this method from this repo https://github.com/hustvl/ViTMatte and it really works good for my case. 

```
def generate_trimap(mask, erode_kernel_size=10, dilate_kernel_size=10):
    erode_kernel = np.ones((erode_kernel_size, erode_kernel_size), np.uint8)
    dilate_kernel = np.ones((dilate_kernel_size, dilate_kernel_size), np.uint8)
    eroded = cv2.erode(mask, erode_kernel, iterations=5)    
    dilated = cv2.dilate(mask, dilate_kernel, iterations=5)
    trimap = np.zeros_like(mask)
    trimap[dilated==255] = 128
    trimap[eroded==255] = 255
    return trimap

# Load VitMatte Model
matte_processor = VitMatteImageProcessor.from_pretrained("hustvl/vitmatte-base-composition-1k")
matte_model = VitMatteForImageMatting.from_pretrained("hustvl/vitmatte-base-composition-1k").to(device)

# mask is the output mask from sam
trimap = generate_trimap(mask, 8, 8)
matte_inputs = matte_processor(images=image_pil.convert("RGB"), trimaps=trimap, return_tensors="pt").to(device)
with torch.no_grad():
        alpha = matte_model(**matte_inputs).alphas[0, 0].to("cpu").numpy()

alpha = (alpha[:H, :W] * 255).astype(np.uint8)
```
